gobblin-core/src/main/java/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java

```diff
@@ -0,0 +1,219 @@
+ *
+ * A prefix and a suffix can be added to the partition, e.g., the partition path can be
+ * 'prefix/2015/11/05/suffix'.
+ *
+ * @author ziliu
+ */
+public abstract class TimeBasedWriterPartitioner<D> implements WriterPartitioner<D> {
+
+  public static final String WRITER_PARTITION_PREFIX = ConfigurationKeys.WRITER_PREFIX + ".partition.prefix";
+  public static final String WRITER_PARTITION_SUFFIX = ConfigurationKeys.WRITER_PREFIX + ".partition.suffix";
+  public static final String WRITER_PARTITION_PATTERN = ConfigurationKeys.WRITER_PREFIX + ".partition.pattern";
+  public static final String WRITER_PARTITION_TIMEZONE = ConfigurationKeys.WRITER_PREFIX + ".partition.timezone";
+  public static final String DEFAULT_WRITER_PARTITION_TIMEZONE = ConfigurationKeys.PST_TIMEZONE_NAME;
+  public static final String WRITER_PARTITION_GRANULARITY = ConfigurationKeys.WRITER_PREFIX + ".partition.granularity";
+  public static final String DEFAULT_WRITER_PARTITION_GRANULARITY = Granularity.HOUR.toString();
+
+  public static final String PARTITIONED_PATH = "partitionedPath";
+  public static final String PREFIX = "prefix";
+  public static final String SUFFIX = "suffix";
+
+  public enum Granularity {
+    YEAR,
+    MONTH,
+    DAY,
+    HOUR,
+    MINUTE;
+  }
+
+  private final String writerPartitionPrefix;
+  private final String writerPartitionSuffix;
+  private final Granularity granularity;
+  private final DateTimeZone timeZone;
+  private final Optional<DateTimeFormatter> timestampToPathFormatter;
+  private final Schema schema;
+
+  public TimeBasedWriterPartitioner(State state, int numBranches, int branchId) {
+    this.writerPartitionPrefix = getWriterPartitionPrefix(state, numBranches, branchId);
+    this.writerPartitionSuffix = getWriterPartitionSuffix(state, numBranches, branchId);
+    this.granularity = getGranularity(state, numBranches, branchId);
+    this.timeZone = getTimeZone(state, numBranches, branchId);
+    this.timestampToPathFormatter = getTimestampToPathFormatter(state, numBranches, branchId);
+    this.schema = getSchema();
+  }
+
+  private String getWriterPartitionPrefix(State state, int numBranches, int branchId) {
+    String propName = ForkOperatorUtils.getPropertyNameForBranch(WRITER_PARTITION_PREFIX, numBranches, branchId);
+    return state.getProp(propName, StringUtils.EMPTY);
+  }
+
+  private String getWriterPartitionSuffix(State state, int numBranches, int branchId) {
+    String propName = ForkOperatorUtils.getPropertyNameForBranch(WRITER_PARTITION_SUFFIX, numBranches, branchId);
+    return state.getProp(propName, StringUtils.EMPTY);
+  }
+
+  private Granularity getGranularity(State state, int numBranches, int branchId) {
+    String propName = ForkOperatorUtils.getPropertyNameForBranch(WRITER_PARTITION_GRANULARITY, numBranches, branchId);
+    String granularityValue = state.getProp(propName, DEFAULT_WRITER_PARTITION_GRANULARITY);
```

Comment:
changed to enum value, but the precondition check is still needed, since the user may specify an invalid value, e.g., "week", which deserves an exception.